osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/alarm/ClearMBusStatusCommandExecutor.java

```diff
@@ -0,0 +1,268 @@
+  private static final ObisCode OBIS_CODE_CLIENT_SETUP_MBUS_3 = new ObisCode("0.3.24.1.0.255");
+  private static final ObisCode OBIS_CODE_CLIENT_SETUP_MBUS_4 = new ObisCode("0.4.24.1.0.255");
+
+  private static final Map<Integer, AttributeAddress> ATTR_READ_STATUS_MAP = new HashMap<>();
+  private static final Map<Integer, AttributeAddress> ATTR_SET_STATUS_MAP = new HashMap<>();
+  private static final Map<Integer, ObisCode> OBIS_CODE_CLIENT_SETUP_MAP = new HashMap<>();
+
+  static {
+    ATTR_READ_STATUS_MAP.put(1, ATTR_READ_STATUS_MBUS_1);
+    ATTR_READ_STATUS_MAP.put(2, ATTR_READ_STATUS_MBUS_2);
+    ATTR_READ_STATUS_MAP.put(3, ATTR_READ_STATUS_MBUS_3);
+    ATTR_READ_STATUS_MAP.put(4, ATTR_READ_STATUS_MBUS_4);
+
+    ATTR_SET_STATUS_MAP.put(1, ATTR_CLEAR_STATUS_MBUS_1);
+    ATTR_SET_STATUS_MAP.put(2, ATTR_CLEAR_STATUS_MBUS_2);
+    ATTR_SET_STATUS_MAP.put(3, ATTR_CLEAR_STATUS_MBUS_3);
+    ATTR_SET_STATUS_MAP.put(4, ATTR_CLEAR_STATUS_MBUS_4);
+
+    OBIS_CODE_CLIENT_SETUP_MAP.put(1, OBIS_CODE_CLIENT_SETUP_MBUS_1);
+    OBIS_CODE_CLIENT_SETUP_MAP.put(2, OBIS_CODE_CLIENT_SETUP_MBUS_2);
+    OBIS_CODE_CLIENT_SETUP_MAP.put(3, OBIS_CODE_CLIENT_SETUP_MBUS_3);
+    OBIS_CODE_CLIENT_SETUP_MAP.put(4, OBIS_CODE_CLIENT_SETUP_MBUS_4);
+  }
+
+  private static final int[] CHANNELS = {1, 2, 3, 4};
+
+  @Autowired
+  public ClearMBusStatusCommandExecutor() {
+    super(ClearMBusStatusRequestDto.class);
+  }
+
+  @Override
+  public ActionResponseDto asBundleResponse(final AccessResultCode executionResult)
+      throws ProtocolAdapterException {
+
+    this.checkAccessResultCode(executionResult);
+
+    return new ActionResponseDto("Clear MBus status was successful");
+  }
+
+  @Override
+  public ClearMBusStatusRequestDto fromBundleRequestInput(final ActionRequestDto bundleInput)
+      throws ProtocolAdapterException {
+
+    this.checkActionRequestType(bundleInput);
+
+    return (ClearMBusStatusRequestDto) bundleInput;
+  }
+
+  @Override
+  public AccessResultCode execute(
+      final DlmsConnectionManager conn,
+      final DlmsDevice device,
+      final ClearMBusStatusRequestDto requestDto,
+      final MessageMetadata messageMetadata)
+      throws ProtocolAdapterException {
+
+    if (Protocol.forDevice(device) != Protocol.SMR_5_1) {
+      throw new NotSupportedByProtocolException("ClearMBusStatus not supported by protocol.");
+    }
+
+    try {
+      for (final int channel : CHANNELS) {
+        this.clearStatus(conn, channel);
+      }
+    } catch (final IOException e) {
+      throw new ConnectionException(e);
+    }
+
+    return AccessResultCode.SUCCESS;
+  }
+
+  private void clearStatus(final DlmsConnectionManager conn, final int channel)
+      throws IOException, ProtocolAdapterException {
+
+    final long statusMask = this.readStatus(conn, channel);
+
+    final AccessResultCode resultCode = this.clearStatus(statusMask, conn, channel);
```

Comment:
In order to reduce network traffic, it would probably be better to skip unnecessary clearStatus and resetAlarm calls to the meter? (By checking the statusMask returned by readStatus)

Reply:
First step is to detect what the current status is
Second step is to update only specific status values
Third step is to invoke a method.
I cannot see how we can do this when the clearStatus and resetAlarm step